lib/pharos/up_command.rb

```diff
@@ -152,42 +141,41 @@ def worker_hosts(config)
       config.hosts.select { |h| h.role == 'worker' }
     end
 
-    # @param master [Pharos::Configuration::Node]
-    # @param features [Pharos::Config]
-    def handle_masters(master, config)
-      log_host_header(master)
-      Phases::ConfigureHost.new(master).call
-      Phases::ConfigureMaster.new(master, config).call
-      Phases::ConfigureClient.new(master).call
-      Phases::ConfigureDNS.new(master, config).call
-      Phases::ConfigureNetwork.new(master, config.network).call
-      Phases::ConfigureMetrics.new(master).call
-      Phases::LabelNode.new(master, master).call
-      Phases::StoreClusterYAML.new(master, config_yaml.read(ENV.to_h)).call
+    def handle_phases(config)
+      all_hosts = config.hosts
+      master_hosts = master_hosts(config)
+      master_host = master_hosts[0]
+      worker_hosts = worker_hosts(config)
+
+      handle_phase(Phases::ValidateHost, all_hosts, ssh: true, parallel: true)
+      handle_phase(Phases::ConfigureHost, all_hosts, ssh: true, parallel: true)
+
+      handle_phase(Phases::ConfigureKubelet, worker_hosts, ssh: true, parallel: true) # TODO: also run this phase in parallel for the master nodes, if not doing an upgrade?
+      
+      handle_phase(Phases::ConfigureMaster, master_hosts, ssh: true, parallel: false)
+      handle_phase(Phases::ConfigureClient, master_hosts, ssh: true, parallel: true)
+
+      # master is now configured and can be used
+      handle_phase(Phases::ConfigureDNS, [master_host], master: master_host)
+      handle_phase(Phases::ConfigureNetwork, [master_host], master: master_host)
+      handle_phase(Phases::ConfigureMetrics, [master_host], master: master_host)
+      handle_phase(Phases::StoreClusterYAML, [master_host], master: master_host, config_content: config_yaml.read(ENV.to_h))
+      handle_phase(Phases::ConfigureBootstrap, [master_host], master: master_host)
+
+      handle_phase(Phases::JoinNode, worker_hosts, ssh: true, parallel: true)
+
+      handle_phase(Phases::LabelNode, all_hosts, master: master_host, ssh: false, parallel: false) # NOTE: uses the @master kube API for each node, not threadsafe
```

Comment:
Assuming that it's fine to label all the nodes at the end, rather than labeling the master nodes before joining the worker nodes.